scrapy/spidermiddlewares/urllength.py

```diff
@@ -14,22 +14,27 @@
 
 class UrlLengthMiddleware:
 
-    def __init__(self, maxlength):
+    def __init__(self, maxlength, stats):
         self.maxlength = maxlength
+        self.stats = stats
 
     @classmethod
-    def from_settings(cls, settings):
+    def from_crawler(cls, crawler):
```

Comment:
I find it hard to figure out how to handle this for backward compatibility (and this is not the first time I face this issue), but I know this is backward incompatible, because [if a subclass defines `from_settings`, that `from_settings` will no longer be called due to this `from_crawler`](https://github.com/scrapy/scrapy/blob/2.4.1/scrapy/utils/misc.py#L145).

Reply:
@Gallaecio we might not inject `stats` instance to a mw, but instead access it by `spider.crawler.stats` so no need to change signatures, and backward compatibility will be kept. 

Is that fine?

Reply:
Sounds great to me, it removes all the backward compatibility headaches :slightly_smiling_face: